torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,332 @@
+    ``scaler`` attempts to use the largest scale factor it can without incurring overflow.
+    It does so by checking the gradients for infs and NaNs during every :meth:`step` or separate :meth:`unscale`.
+    If no infs/NaNs are found, :meth:`step` runs the underlying ``optimizer.step()`` as usual and
+    :meth:`update` multiplies the scale factor by the growth factor.  If infs/NaNs are found, :meth:`step` skips the
+    underlying ``optimizer.step()`` (so the params themselves remain unpolluted) and multiplies the scale factor by
+    the backoff factor.
+
+    The scale factor often causes infs/NaNs to appear in gradients for the first few iterations as its
+    value calibrates.  ``scaler.step`` will skip the underlying ``optimizer.step()`` for these
+    iterations.  After that, step skipping should occur rarely (once every few hundred iterations).
+
+    Arguments:
+        init_scale (float, optional, default=2.**24):  Initial scale factor.
+        growth_factor (float, optional, default=1.001):  Factor by which the scale is multiplied during
+            :meth:`update` if no inf/NaN gradients were found this iteration.
+        backoff_factor (float, optional, default=0.5):  Factor by which the scale is multiplied during
+            :meth:`update` if inf/NaN gradients were found this iteration.
+        enabled (bool,optional, default=True):  If ``False``, disables gradient scaling. :meth:`step` simply
+            invokes the underlying ``optimizer.step()``, and other methods become no-ops.
+    """
+    def __init__(self,
+                 init_scale=2.**24,
+                 growth_factor=1.001,
+                 backoff_factor=0.5,
+                 enabled=True):
+        self._enabled = enabled
+        if enabled:
+            self._scale = torch.full((1,), init_scale, dtype=torch.float32, device="cuda")
+            self._growth_factor = growth_factor
+            self._backoff_factor = backoff_factor
+            self._per_optimizer_states = defaultdict(dict)
+
+    def scale(self, outputs):
+        """
+        Multiplies ('scales') a tensor or list of tensors by the scale factor.
+
+        These tensors are typically the outputs of a network. When ``backward()``
+        is called on scaled outputs the resulting gradients will also be scaled.
+
+        Arguments:
+            outputs (Tensor or iterable of Tensors):  Outputs to scale.
+
+        Returns:
+            Scaled outputs.  If this instance of :class:`AmpScaler` is not enabled, outputs are returned unmodified.
+        """
+        if not self._enabled:
+            return outputs
+
+        # Short-circuit for the common case.
+        if isinstance(outputs, torch.Tensor):
+            assert outputs.is_cuda
+            return outputs * self._scale.to(outputs.device)
+
+        # Invoke the more complex machinery only if we're treating multiple outputs.
+        per_device_scale = _PerDeviceHelper(self._scale)
```

Comment:
Using the helper class seems unecessary. You can just cache the .to's directly in a dict, right?